### PR TITLE
fix(telemetry): correct create counter, send-failure handling, and aggregation accuracy

### LINKED ISSUE #1874: Telemetry: session_creates_since_last_snapshot is always 0, create-trend field carries no signal
**Note:** Claude handled the investigation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

Follow-up to #1863 (opt-in usage telemetry). Directly affects the premise of #1870.

## Problem

`usage_snapshot.session_creates_since_last_snapshot` is a wire field that is **always `0`**. Both emit sites pass a literal `0`; nothing ever counts session creates. So the field ships in every snapshot but carries no signal.

This matters beyond a dead field: #1870 ("time-aggregated usage snapshots") explicitly rests on it, stating "Total volume is already preserved by `session_creates_since_last_snapshot`, so what is lost is the agent/model attribution...". That premise is false today. Short-lived sessions that start and end between two 12h ticks are invisible in the status/agent/model maps AND uncounted by the create trend, so total volume is **not** preserved.

## What's there today

The TUI passes `0` hardcoded:

```rust
// src/tui/app.rs:1320-1328
fn build_telemetry_snapshot(&self) -> Option<crate::telemetry::UsageSnapshot> {
    crate::telemetry::build_usage_snapshot(
        crate::telemetry::Surface::Tui,
        self.home.instances(),
        false,
        false,
        0,   // session_creates_since_last_snapshot
    )
}
```

So does serve:

```rust
// src/server/mod.rs:1663-1675
async fn build_serve_snapshot(state: &AppState) -> Option<...> {
    ...
    crate::telemetry::build_usage_snapshot(
        crate::telemetry::Surface::Serve,
        &instances,
        web_seen,
        cockpit_seen,
        0,   // "left at 0 for v1"
    )
}
```

The field is faithfully serialized (`src/telemetry/events.rs:92-94`) and documented as a trend counter, but no producer increments it.

## Root cause

`build_usage_snapshot` is stateless and takes the count as a parameter (`src/telemetry/mod.rs:132-138`). There is no accumulator anywhere that observes session-create events and feeds a running count to the next snapshot, so every caller has nothing to pass but `0`.

## Proposal

A (minimum): Wire a process-local create counter that increments whenever a session is created and resets to its value at each snapshot flush.
- Serve: increment in the session-create path that `aoe serve` owns, read-and-reset in `build_serve_snapshot` (`server/mod.rs:1663`), mirroring the existing `telemetry_web_seen` / `telemetry_cockpit_seen` swap-on-read pattern (`server/mod.rs:1665-1666`).
- TUI: increment on create in the home loop, read-and-reset in `build_telemetry_snapshot` (`app.rs:1320`).

B (alternative): If the field is not wanted in v1, remove it from the schema rather than ship an always-zero field (AGENTS.md: no dead fields). This is the cheaper path if create-trend is deferred, but it conflicts with #1870's design, so coordinate.

A is preferred because #1870 already depends on the field carrying real data.

## Why now

- #1870's stated design premise is currently false; fixing this is a prerequisite for that work to mean what it claims.
- Always-zero wire fields violate the no-dead-code rule in AGENTS.md.

## User stories / regression tests

Implementation MUST add tests covering each story below; PR is not done without them.

- As a telemetry consumer, when N sessions are created during a snapshot window, then the next `usage_snapshot.session_creates_since_last_snapshot` equals N.
  - Suggested test layer: Rust unit/integration around the accumulator (`tests/telemetry.rs`).
- As a telemetry consumer, when a snapshot is emitted, then the create counter resets so the following snapshot starts from 0.
  - Suggested test layer: Rust unit (read-and-reset semantics).
- As an opted-out user, when sessions are created, then no counter state is sent (default-off path unchanged).
  - Suggested test layer: Rust integration (`tests/telemetry.rs`).

## Open questions

- Should a session create be counted at the substrate the surface owns (serve daemon create path) or at a shared session-storage layer so both TUI and serve observe the same events without double counting?
- For the TUI, which is short-lived, is a per-process counter even worth it, or should this be serve-only (matching #1870's "almost entirely a serve win")?

## Out of scope

- The broader 30-min local sampling / peak-concurrency aggregation in #1870. This issue only makes the existing create-trend field carry a real value (or removes it).
- Any schema-version bump unless option B (removal) is chosen.

## Cross-references

- #1863 — PR that introduced the always-zero field.
- #1870 — depends on this field being populated; its "total volume is already preserved" claim is blocked on this fix.
- #1762 — telemetry tracking plan.

### LINKED ISSUE #1875: Telemetry: signals silently lost on send failure (web_seen/cockpit_seen reset and CLI slot claimed before POST)
**Note:** Claude handled the investigation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

Follow-up to #1863 (opt-in usage telemetry). Same root pattern as #1874.

## Problem

Two telemetry signals are consumed (state reset / slot claimed) *before* the corresponding POST is confirmed. Sends are fire-and-forget and swallow every error with no retry (`src/telemetry/mod.rs:222-245`, `docs/telemetry.md:85`), so a failed send silently discards the signal for that whole window. The signal is gone, not retried.

## What's there today

**A. `web_seen` / `cockpit_seen` swapped to false at build time.**

```rust
// src/server/mod.rs:1663-1666
async fn build_serve_snapshot(state: &AppState) -> Option<...> {
    let web_seen = state.telemetry_web_seen.swap(false, Ordering::Relaxed);
    let cockpit_seen = state.telemetry_cockpit_seen.swap(false, Ordering::Relaxed);
    ...
}
```

The flags are reset to false the instant the snapshot is built. The snapshot then goes out via `spawn_snapshot` (`server/mod.rs:1651`), which posts detached and swallows failure. If the POST fails, "the dashboard / cockpit was opened this window" is lost; the next snapshot reports false.

**B. CLI `process_start` daily slot claimed before send.**

```rust
// src/telemetry/mod.rs:270-274
pub async fn flush_cli_process_start() {
    if is_opted_in() && state::claim_cli_process_start_slot(CLI_PROCESS_START_MIN_GAP) {
        flush_process_start(Surface::Cli).await;
    }
}
```

`claim_cli_process_start_slot` stamps "now" and returns true (`state.rs:117`) *before* `flush_process_start` runs. If that send fails (`mod.rs:244` swallows it), the daily slot is already consumed, so no CLI `process_start` lands for the rest of the day, with no retry.

## Root cause

Both consume local state optimistically, assuming the send succeeds, but the send path is best-effort with no success signal back to the caller. There is no "restore on failure" because `spawn_snapshot` / `flush_process_start` return nothing useful and run detached.

## Proposal

A (preferred, minimal): Defer the state mutation until the send is confirmed.
- `process_start`: make `flush_process_start` report success, and only `claim_cli_process_start_slot` (stamp the file) when the POST actually succeeded. A failed send leaves the slot open so the next invocation retries.
- `web_seen` / `cockpit_seen`: read the flags without resetting, send, and reset only on a confirmed 2xx. On failure leave them set so the next snapshot carries them.

B (cheaper, explicit): Accept the loss but document it. Keep the current optimistic reset and add a one-line note in `docs/telemetry.md` "Failure isolation" that a dropped send also drops that window's `web_seen` / `cockpit_seen` and the day's CLI `process_start`. Only acceptable if we decide best-effort coarse signals are fine to lose.

A is preferred for the CLI slot (losing a whole day on one transient failure is a real undercount); B may be fine for the seen flags given they re-arm the next time the UI is opened.

## User stories / regression tests

Implementation MUST add tests covering each story below; PR is not done without them.

- As a telemetry consumer, when the snapshot POST fails, then `web_seen` / `cockpit_seen` remain set so the next snapshot still reports them.
  - Suggested test layer: Rust unit/integration around `build_serve_snapshot` with a failing endpoint (`tests/telemetry.rs`).
- As a telemetry consumer, when a CLI `process_start` send fails, then the daily throttle slot is not consumed and the next invocation attempts again.
  - Suggested test layer: Rust integration (`tests/telemetry.rs`, unreachable-endpoint pattern already exists).
- As an opted-out user, when any send path runs, then no slot is claimed and no flag state changes.
  - Suggested test layer: Rust integration (default-off path).

## Open questions

- Should the seen flags use option B (accept loss, re-arm next open) while the CLI slot uses option A (defer claim)? The asymmetry is defensible given their different cost-of-loss.
- Confirming a 2xx requires threading the response status out of `post` (currently it only logs). Acceptable surface change?

## Out of scope

- Adding retries or offline buffering (depends on #1873's idempotency key first).
- The always-zero create counter (#1874).

## Cross-references

- #1863 — PR that introduced both signals.
- #1874 — same "signal lost / never populated" family.
- #1873 — idempotency key; prerequisite before any retry-based fix.

### LINKED ISSUE #1876: Telemetry: aggregation accuracy, model_bucket substring false-positives and features reflect global not active profile
**Note:** Claude handled the investigation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

Follow-up to #1863 (opt-in usage telemetry).

## Problem

Two telemetry aggregation values are computed in ways that skew the resulting counts. Neither leaks content (both stay within the closed allowlist); they just mis-bucket, so the maintainer-facing numbers are wrong.

## A. `model_bucket` short substring needles false-positive

```rust
// src/telemetry/sanitize.rs:43-58
const FAMILIES: &[(&str, &[&str])] = &[
    ("claude", &["claude", "sonnet", "opus", "haiku"]),
    ("openai", &["gpt", "openai", "codex", "o1", "o3", "o4"]),
    ...
];
for (family, needles) in FAMILIES {
    if needles.iter().any(|n| lower.contains(n)) {   // substring match
        return family;
    }
}
```

The openai needles include `"o1"`, `"o3"`, `"o4"`, matched with `lower.contains(n)`. Any model string that merely *contains* `o1` / `o3` / `o4` as a substring buckets as `openai` (for example a name like `macro1-...`, `kilo3`, or anything with those two chars adjacent). The tokens are too short for a substring test, so `sessions_by_model_bucket` over-counts `openai`.

## B. `features` reflects global config, not the active profile

```rust
// src/telemetry/features.rs:22-31
pub fn active_features(config: &Config) -> BTreeMap<String, bool> {
    features.insert("worktree".to_string(), config.worktree.enabled);
    features.insert("sandbox".to_string(), config.sandbox.enabled_by_default);
    features.insert("cockpit".to_string(), config.cockpit.enabled);
    ...
}
```

Called with `Config::load_or_warn()` (`src/telemetry/mod.rs:143`), i.e. the global config. A user who runs a profile with worktree / sandbox / cockpit overrides reports the *global* feature state, not what they actually run, so feature-adoption numbers skew for profile users.

## Proposal

A (model bucket): match the short tokens on a boundary instead of bare substring. Either anchor (`starts_with` / regex `\bo[134]\b`) or require a delimiter, so `o3` matches `o3-mini` but not `kilo3`. Keep `claude` / `gpt` / `gemini` etc. as substrings; only the 2-char tokens need tightening. Add table tests for the false-positive cases.

B (features): decide the intended semantic and document it. Either (i) report per-active-profile effective config (resolve the profile before reading the flags), or (ii) keep global but rename / document the field as "global default feature state" so the number is not misread as per-session adoption. (i) is more useful if profiles are common; (ii) is a one-line doc fix if global-default is actually what we want.

## User stories / regression tests

Implementation MUST add tests covering each story below; PR is not done without them.

- As a telemetry consumer, when a model string contains `o1`/`o3`/`o4` only as an incidental substring, then it does not bucket as `openai`.
  - Suggested test layer: Rust unit (`src/telemetry/sanitize.rs` tests, add false-positive cases).
- As a telemetry consumer, when a real OpenAI model (`o3-mini`, `gpt-5`, `codex`) is seen, then it still buckets as `openai`.
  - Suggested test layer: Rust unit (existing `model_buckets_map_to_families`).
- As a telemetry consumer, when a user runs a profile that overrides a tracked feature, then the reported `features` map reflects the agreed semantic (per chosen option A.i/A.ii).
  - Suggested test layer: Rust unit around `active_features` with a profile override.

## Open questions

- For features: is the intent install-level default adoption (global is correct) or actual-usage adoption (profile-resolved is correct)? This decides whether B is a code fix or a doc fix.

## Out of scope

- Any change to the sanitizer's privacy guarantees; both buckets already stay within the allowlist.

## Cross-references

- #1863 — PR that introduced the sanitizer and feature registry.

### PR DESCRIPTION
**Pull request by Seluj78**
**Note:** Claude handled the implementation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Description

Three linked telemetry data-correctness fixes, all follow-ups to the opt-in telemetry in #1863. Each made a shipped wire field carry wrong or no signal; this PR makes them truthful.

**#1874 - create-trend counter was always 0.** `usage_snapshot.session_creates_since_last_snapshot` was a wire field that every producer passed a literal `0` for, so it carried no signal (and #1870's design rests on it). `aoe serve` now keeps a process-local create counter, incremented in the `create_session` handler, and the snapshot reports it. Per the issue's open question this is serve-only for now (the TUI is short-lived and #1870 frames it as "almost entirely a serve win"); the TUI counter is tracked as a follow-up in #1897.

**#1875 - signals silently lost on send failure.** Sends were fire-and-forget and treated any transport `Ok` (including a 4xx/5xx rejection at the gateway) as success, while consuming local state before the POST was confirmed. A failed send permanently dropped that window's `web_seen` / `cockpit_seen` and the day's CLI `process_start`. Now `post()` returns whether delivery was confirmed (transport `Ok` and a 2xx), and state is consumed only on a confirmed send: the CLI daily slot is claimed only on success (failed sends retry, bounded to once per hour), and the serve seen-flags plus the new create counter are cleared only after a confirmed snapshot. The same fix corrects an adjacent bug surfaced in review: the dedup fingerprint was recorded before the send, so a failed boot send could poison the cache into dropping a later identical retry.

**#1876 - aggregation accuracy.** `model_bucket` matched the 2-char OpenAI tokens `o1`/`o3`/`o4` as bare substrings, so names like `kilo3` or `macro1` over-counted as `openai`; they now require a whole-token boundary match while distinctive needles stay substring matches. The `features` map is documented as the global install default (pre-profile-merge), an install-level adoption signal, rather than being silently misread as per-session usage.

Closes #1874
Closes #1875
Closes #1876

## PR Type

- [ ] New Feature
- [x] Bug Fix
- [ ] Refactor
- [ ] Documentation
- [ ] Infrastructure / CI

## How to test

- [ ] Point `AOE_TELEMETRY_ENDPOINT` at a local sink, opt in, and run `aoe serve`; create a session via the dashboard and confirm the next `usage_snapshot` reports `session_creates_since_last_snapshot` > 0, then `0` again on the following snapshot.
- [ ] With the local sink returning a 5xx, opt in and run `aoe <subcommand>` twice within a day; confirm the second invocation still attempts a CLI `process_start` (the daily slot was not consumed by the failed send).
- [ ] With the local sink up again, confirm the daily CLI `process_start` slot is then claimed and a same-day re-run does not re-send.
- [ ] Open the web dashboard against a failing sink; confirm `web_seen` stays set and is reported again on the next snapshot rather than being lost.
- [ ] Confirm a model name like `kilo3` buckets as `other` and `o3-mini` buckets as `openai` in the emitted snapshot.

## Checklist

- [x] New and existing tests pass
- [x] Documentation was updated where necessary
- [ ] For UI changes: included screenshot or recording

## Test Coverage Analysis

**Web dashboard / cockpit (Playwright user story)**
- [x] N/A: this PR doesn't change a user-facing dashboard flow

**TUI / CLI (e2e test)**
- [ ] N/A: this PR doesn't change TUI rendering, a CLI subcommand, or session lifecycle
- [x] Skipped a test, because: the changes are backend telemetry only (no TUI rendering or CLI subcommand behavior change). Regression coverage lives in Rust unit/integration tests: `tests/telemetry.rs` (create count carried, failed send leaves the CLI slot open), `src/server/mod.rs` (create-counter decrement preserves concurrent creates), and `src/telemetry/sanitize.rs` (model-bucket false-positive table).

## AI Usage

- [x] AI was used

**AI Model/Tool used:** Claude (via Claude Code, agent-of-empires `/aoe-implement` skill)

**Any Additional AI Details you'd like to share:**
Investigation, plan (with a multi-model `/debate` consult), and implementation drafted by Claude under my direction. I made the design calls (serve-only counter, keep-global features semantic), reviewed each commit, and own the test steps above.

- [x] I am an AI Agent filling out this form (check box if true)

# Telemetry Correctness Fixes

## What changed (high-level)
This PR implements three linked fixes that make telemetry data more accurate and resilient:

- Session create counter: serve now tracks sessions created since the last telemetry snapshot and reports/reset that count (TUI still reports 0 for now).
- Send semantics and retry behavior: telemetry is only finalized/consumed after confirmed delivery (HTTP 2xx). Failed sends no longer permanently drop signals; CLI daily process_start claims are only recorded after success and retries are bounded.
- Model bucketing and feature semantics: short OpenAI tokens (o1/o3/o4) are matched only at whole-token boundaries to avoid false positives; the `features` map is documented/treated as the install-wide default (pre-profile-merge), not per-session active-profile values.

Also includes documentation updates and tests for all three areas.

## Benefits
- More reliable telemetry (prevents silent loss on send failures).
- Accurate session-create counts in serve snapshots.
- Fewer false positives in model attribution.
- Clearer semantics for feature reporting and better-controlled retry/backoff behavior.

## Notable code-level changes
- serve: added a process-local atomic counter for session creates and converted seen flags to monotonic AtomicU32 counters; introduced telemetry_last_reported to defer clearing until confirmed sends.
- telemetry sending: post/send now returns delivery outcome; new SendOutcome enum (Sent, Deduped, Failed); dedup fingerprints and state mutations are recorded only after confirmed sends.
- CLI throttle state: replaced single claim API with cli_process_start_due(success_gap, retry_gap) and record_cli_process_start(success) to separate "is due" checks from "record result" and bound retry behavior.
- sanitization: introduced a Needle matcher supporting substring and whole-token matches; model_bucket uses token matches for short OpenAI tokens.
- tests & docs: updated telemetry docs and added tests covering session counts, send-failure behavior and retry semantics, and model bucketing edge cases.

## Potential follow-ups / considerations
- TUI session-create counter tracked separately (`#1897`); consider consolidating serve/TUI reporting for consistency.
- Review backoff parameters and observability for repeated send failures to ensure visibility without excessive retries.

### CODE REVIEW

File: src/server/mod.rs [hunk 1]
```diff
@@ -1642,36 +1655,97 @@ fn spawn_telemetry_loop(state: Arc<AppState>) {
...
+    };
+    if outcome != crate::telemetry::SendOutcome::Sent {
+        return;
+    }
+    if reported.web_seen {
+        state.telemetry_web_seen.store(false, Ordering::Relaxed);
+    }
+    if reported.cockpit_seen {
+        state.telemetry_cockpit_seen.store(false, Ordering::Relaxed);
+    }
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply src/server/mod.rs]: Good catch, fixed in https://github.com/agent-of-empires/agent-of-empires/pull/1898/commits/f36f2d2969924b4a1087949909e47bc5db5ebaf4. web_seen / cockpit_seen are now AtomicU32 open counts cleared by subtracting exactly the reported value (the same path the create counter uses via decrement_reported_count), so an open that lands during the in-flight send is preserved for the next snapshot. The wire field stays the boolean count > 0.

File: src/telemetry/mod.rs [hunk 2]
```diff
@@ -240,37 +251,61 @@ async fn post<T: serde::Serialize>(event: &T) {
...
-    if is_opted_in() && state::claim_cli_process_start_slot(CLI_PROCESS_START_MIN_GAP) {
-        flush_process_start(Surface::Cli).await;
+    if !is_opted_in() {
+        return;
+    }
+    if !cli_process_start_due(CLI_PROCESS_START_MIN_GAP, CLI_PROCESS_START_RETRY_GAP) {
+        return;
     }
+    let confirmed = flush_process_start(Surface::Cli).await;
+    record_cli_process_start(confirmed);
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply src/telemetry/mod.rs:308]: Leaving this as-is for now. The cross-process race predates this PR: the previous claim_cli_process_start_slot was also read-file-then-write-file with no lock, so two concurrent aoe invocations could already both claim. The cost is bounded (a few duplicate once-per-day process_start events when a user launches multiple aoe within the same ~2s window), and the proper fix (a file-locked claim/finalize two-phase) is a separate change that must avoid re-introducing the slot-consumed-on-failure bug this PR fixes. Tracking it separately rather than expanding this PR's scope.

**Review by njbrake — APPROVED** [review]
